src/bp/invoice.rs

```diff
@@ -0,0 +1,128 @@
+// LNP/BP Core Library implementing LNPBP specifications & standards
+// Written in 2020 by
+//     Dr. Maxim Orlovsky <orlovsky@pandoracore.com>
+//
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the MIT License
+// along with this software.
+// If not, see <https://opensource.org/licenses/MIT>.
+
+use bitcoin::hashes::sha256d;
+use bitcoin::secp256k1::Signature;
+use bitcoin::Address;
+use miniscript::{descriptor::DescriptorPublicKey, Descriptor, Miniscript};
+use url::Url;
+
+use crate::bp::blind::OutpointHash;
+use crate::bp::chain::AssetId;
+use crate::bp::{Chain, HashLock, P2pNetworkId, Psbt, ScriptPubkeyFormat};
+use crate::lnp::{tlv, Features};
+use crate::secp256k1;
+
+#[derive(Tlv)]
+#[lnpbp_crate(crate)]
+pub enum FieldType {
+    #[tlv(type = 0x01)]
+    Payers,
+}
+
+#[derive(Lnp)]
+#[tlv_types(FieldType)]
+#[lnpbp_crate(crate)]
+pub struct Invoice {
+    network: P2pNetworkId,
+
+    /// List of beneficiary dests ordered in most desirable first order
+    beneficiaries: Vec<Beneficiary>,
+
+    /// Optional list of payers authored to pay
+    #[tlv(type = FieldType::Payers)]
+    payers: Vec<Payer>,
+    quantity: Quantity,
+    price: Option<AmountExt>,
+
+    /// If the price of the asset provided by fiat provider URL goes below this
+    /// limit the merchant will not accept the payment and it will become
+    /// expired
+    fiat_requirement: Option<Fiat>,
+    merchant: Option<String>,
+    asset: Option<AssetId>,
+    purpose: Option<String>,
+    details: Option<Details>,
+    expiry: Option<i64>,
+
+    #[tlv_unknown]
+    unknown: Vec<tlv::Map>,
+    signature: Option<Signature>,
+}
+
+#[non_exhaustive]
+pub enum Beneficiary {
```

Comment:
Does it make sense to have "push channel opening" as possible beneficiary?
Something like "pay me on lightning if you find a path or open a channel while pushing the amount otherwise". Technically payer already has the data he requires in order to open a channel (which is basically payee's nodeid + IP that he can get from gossip protocol). What's missing is the information that the invoice will be considered "payed" in case of a proper "push channel opening".

Reply:
I was thinking about adding such flag. But it seems that it does not make a lot of sense, since it will be up to the payee to decide whether to open direct channel or route the payment.

Basically with the current structure you are still able to create "open channel with me" type of invoices by just leaving amount equal to 0 and specifying lightning payment option with node id/ip address - and that's all.